src/http/notification/NotificationSubscriptionHttpHandler.ts

```diff
@@ -0,0 +1,127 @@
+    private readonly source: EventEmitter,
+    subscriptionHandlers: SubscriptionHandler<Subscription>[],
+  ) {
+    super();
+    for (const handler of subscriptionHandlers) {
+      this.subscriptionHandlers.set(handler.getType(), handler);
+    }
+    this.source.on(AS.Create, async(id: ResourceIdentifier): Promise<void> => this.onResourceChanged(id, AS.Create));
+    this.source.on(AS.Update, async(id: ResourceIdentifier): Promise<void> => this.onResourceChanged(id, AS.Update));
+    this.source.on(AS.Delete, async(id: ResourceIdentifier): Promise<void> => this.onResourceChanged(id, AS.Delete));
+  }
+
+  public getSupportedTypes(): string[] {
+    return [ ...this.subscriptionHandlers.keys() ];
+  }
+
+  public async canHandle({ operation }: OperationHandlerInput): Promise<void> {
+    if (operation.method !== 'POST') {
+      throw new NotImplementedHttpError('This handler only supports POST operations');
+    }
+  }
+
+  public async handle({ request }: OperationHttpHandlerInput): Promise<ResponseDescription> {
+    const body = await request.read();
+    if (!body) {
+      throw new BadRequestHttpError('A body should be provided');
+    }
+    const subscriptionRequest = JSON.parse(body.toString());
+    if (!subscriptionRequest.topic || !subscriptionRequest.target) {
+      throw new BadRequestHttpError('A `topic` and `target` should be provided in the body');
+    }
+    const subscriptionType: string = subscriptionRequest.type;
+
+    // Check if the requested subscription type is supported/available
+    const subscriptionHandler = this.subscriptionHandlers.get(subscriptionType);
+    if (!subscriptionHandler) {
+      throw new BadRequestHttpError(
+        `Subscription type "${subscriptionType}" not supported`,
+      );
+    }
+
+    // Check permissions for the topic
+    const topicURI: string = subscriptionRequest.topic;
+    const credentials = await this.credentialsExtractor.handleSafe(request);
+    if (!credentials.agent?.webId) {
+      throw new BadRequestHttpError('No WebId present in request');
+    }
+    const topicIdentifier = { path: topicURI };
+    const permissionMap: PermissionMap = await this.permissionReader.handleSafe({
+      credentials,
+      requestedModes: new IdentifierSetMultiMap<AccessMode>([
+        [ topicIdentifier, AccessMode.read ],
+      ]),
+    });
+    const permissionSet = permissionMap.get(topicIdentifier);
+    if (!permissionSet?.public?.read && !permissionSet?.agent?.read) {
+      throw new BadRequestHttpError('Agent not allowed to read resource.');
+    }
+
+    const existingTopic = await this.notificationStorage.get(topicURI);
+    const topic: Topic = existingTopic ?? { subscriptions: {}};
+
+    // Create and save the new subscription
+    const subscription: Subscription = subscriptionHandler.subscribe(subscriptionRequest);
+    topic.subscriptions[credentials.agent.webId] = subscription;
+    await this.notificationStorage.set(topicURI, topic);
```

Comment:
@joachimvh Do you have some advice/tip on this issue? Maybe a small code snippet that showcases how to acquire & release a lock on the `topicURI` resource in the `notificationStorage`?

Reply:
I don't have the context here so I can't comment on what the best solution is, but the locking store is the best place to look at how locking can be used: https://github.com/CommunitySolidServer/CommunitySolidServer/blob/d7d1480e8e44f9687059907d429a0024bca7824f/src/storage/LockingResourceStore.ts#L50-L51

So you just need an `(Expiring)ReadWriteLocker` to which you can pass a function that needs to be executed in a lock.

As an aside, seeing so many files added/changed at the same time without having seen a discussion about how to best solve this worries me a bit as this means the discussion might happen after the code is already written, which could cause additional work and refactors if there are disagreements.

Reply:
This PR has kind of evolved over time (with some discussion/comments) through pull requests:  https://github.com/CommunitySolidServer/CommunitySolidServer/pull/1080 and https://github.com/CommunitySolidServer/CommunitySolidServer/pull/1074 

We can of course have a chat discussing this change, either here in this PR (or separately if desired?) 👍

Reply:
I actually went back to those PRs to check but noticed there was also not that much discussion there.

I just want to make sure that nobody puts in work that has to be reverted and feels like they wasted time. So I also want to hear what you think is the most preferable way for you to do this. E.g., I can already go through the PR once even though it's not finished, we can have a call to discuss some aspects (perhaps after the initial overview), etc.

Reply:
Yeah, I think at this point, this initial implementation of webhook notifications have been done and then cleaned up/refactored into a state that we (as consumers/clients) are quite happy with.

However, the implementation it might not yet perfectly fit with the design/expectations of this project, but I do believe that we can resolve any differences here within this PR. A couple of review iterations might be necessary with you(the owners/maintainers of this repo) to reach a satisfactory result.

Also, I think a separate discussion would only be necessary if the identified issues/differences with this PR are so vast it can't productively be dealt with/handled here within this PR.

That said, I think it's time for a critical review by the maintainers of this repo. 🙂

Reply:
I will have a look later then